File: sample-textcapture/TextCapture/RTRSelectedAreaView.m
// ABBYY® Real-Time Recognition SDK 1 © 2016 ABBYY Production LLC.
// ABBYY is either a registered trademark or a trademark of ABBYY Software Ltd.

#import "RTRSelectedAreaView.h"

#pragma mark - Constants

/// The area of interest border thickness.
static const CGFloat AreaBorderThickness = 1.f;

/// The background color for the rest of the screen outside the area of interest.
static const CGFloat CropFogRedColorComponent = 0.f;
static const CGFloat CropFogGreenColorComponent = 0.f;
static const CGFloat CropFogBlueColorComponent = 0.f;
static const CGFloat CropFogAlphaColorComponent = 0.7f;

/// The area of interest border color.
static const CGFloat CropBorderRedColorComponent = 1.f;
static const CGFloat CropBorderGreenColorComponent = 1.f;
static const CGFloat CropBorderBlueColorComponent = 1.f;
static const CGFloat CropBorderAlphaColorComponent = 0.5f;

@implementation RTRSelectedAreaView {
	/// Background color.
	UIColor* _areaFogColor;
	/// Border color.
	UIColor* _areaBorderColor;
}

- (id)initWithCoder:(NSCoder*)aDecoder
{
	self = [super initWithCoder:aDecoder];

	if(self != nil) {
		[self doInit];
	}
	
	return self;
}

- (id)initWithFrame:(CGRect)frame
{
	self = [super initWithFrame:frame];

	if(self != nil) {
		[self doInit];
	}

	return self;
}

- (void)doInit
{
	_areaFogColor = [UIColor colorWithRed:CropFogRedColorComponent green:CropFogGreenColorComponent
		blue:CropFogBlueColorComponent alpha:CropFogAlphaColorComponent];

	_areaBorderColor = [UIColor colorWithRed:CropBorderRedColorComponent green:
		CropBorderGreenColorComponent blue:CropBorderBlueColorComponent alpha:CropBorderAlphaColorComponent];

	self.exclusiveTouch = YES;
}

- (void)setSelectedArea:(CGRect)selectedArea
{
	_selectedArea = selectedArea;
	[self setNeedsDisplay];
}

- (void)drawRect:(CGRect)rect
{
	[super drawRect:rect];

	CGContextRef currentContext = UIGraphicsGetCurrentContext();

	CGContextSaveGState(currentContext);
	CGContextTranslateCTM(currentContext, 0, 0);

	[self drawFogLayerInContext:currentContext];
	[self drawBorderLayerInContext:currentContext];

	CGContextRestoreGState(currentContext);
}

- (void)drawFogLayerInContext:(CGContextRef)context
{
	CGContextSaveGState(context);

	CGRect scaledBounds = self.superview.bounds;

	CGContextAddRect(context, scaledBounds);
	[self addPathForSelectedAreaToContext:context];
	CGContextEOClip(context);

	// Fill the background outside the area of interest
	CGContextSetFillColorWithColor(context, _areaFogColor.CGColor);
	CGContextFillRect(context, scaledBounds);

	CGContextRestoreGState(context);
}

- (void)drawBorderLayerInContext:(CGContextRef)context
{
	// Draw the border of the area of interest
	[self addPathForSelectedAreaToContext:context];
	CGContextSetStrokeColorWithColor(context, _areaBorderColor.CGColor);
	CGContextSetLineWidth(context, AreaBorderThickness);
	CGContextDrawPath(context, kCGPathStroke);
}

- (void)addPathForSelectedAreaToContext:(CGContextRef)context
{
	CGPoint origin = self.selectedArea.origin;
	CGFloat width = CGRectGetWidth(self.selectedArea);
	CGFloat height = CGRectGetHeight(self.selectedArea);
	CGPoint points[] = {
		origin,
		CGPointMake(self.selectedArea.origin.x + width, origin.y),
		CGPointMake(self.selectedArea.origin.x + width, origin.y + height),
		CGPointMake(self.selectedArea.origin.x, origin.y + height)
	};

	CGContextAddLines(context, points, 4);
	CGContextClosePath(context);
}

@end

File: sample-textcapture-swift/TextCapture/RTRSelectedAreaView.swift
// ABBYY® Real-Time Recognition SDK 1 © 2016 ABBYY Production LLC.
// ABBYY is either a registered trademark or a trademark of ABBYY Software Ltd.

import UIKit

class RTRSelectedAreaView: UIView {
	/// Border thickness of capture zone
	private let AreaBorderThickness : CGFloat = 1.0
	/// Background color
	private let AreaFogColor: UIColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.7)
	/// Border color of capture zone
	private let AreaBorderColor: UIColor = UIColor(red: 1, green: 1, blue: 1, alpha: 0.5)

	internal var selectedArea: CGRect = CGRect.zero {
		didSet {
			weak var weakSelf = self
			DispatchQueue.main.async {
				if let strongSelf = weakSelf {
					strongSelf.setNeedsDisplay()
				}
			}
		}
	}

//# MARK: - LifeCycle

	override init(frame: CGRect) {
		super.init(frame: frame)
		doInit()
	}

	required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
		doInit()
	}

//# MARK: - Private

	private func doInit() {
		isExclusiveTouch = true
	}

	override func draw(_ rect: CGRect) {
		super.draw(rect)

		if let currentContext = UIGraphicsGetCurrentContext() {
			currentContext.saveGState()
			currentContext.translateBy(x: 0, y: 0)

			drawFogLayer(currentContext)
			drawBorderLayer(currentContext)

			currentContext.restoreGState()
		}
	}

	private func drawFogLayer(_ context: CGContext) {
		context.saveGState()

		if let superview = self.superview {
			let scaledBounds = superview.bounds

			// Fill the background
			context.setFillColor(AreaFogColor.cgColor)
			context.fill(scaledBounds)

			let intersection = selectedArea.intersection(scaledBounds)
			context.addRect(intersection)
			context.clip()
			context.clear(intersection)

			context.setFillColor(UIColor.clear.cgColor)
			context.fill(intersection)

			context.restoreGState()
		}
	}

	private func drawBorderLayer(_ context: CGContext) {
		// Draw the outline of the capture zone
		addPathForSelectedArea(context)
		context.setStrokeColor(AreaBorderColor.cgColor) 
		context.setLineWidth(AreaBorderThickness) 
		context.drawPath(using: CGPathDrawingMode.stroke) 
	}

	private func addPathForSelectedArea(_ context: CGContext) {
		let origin = selectedArea.origin
		let width = selectedArea.width
		let height = selectedArea.height

		let points = [origin,
			CGPoint.init(x: selectedArea.origin.x + width, y: origin.y),
			CGPoint.init(x: selectedArea.origin.x + width, y: origin.y + height),
			CGPoint.init(x: selectedArea.origin.x, y: origin.y + height)]

		context.addLines(between: points) 
		context.closePath() 
	}
}
